quixstreams/topic_manager.py

```diff
@@ -0,0 +1,770 @@
+
+    class TopicValidationError(Exception):
+        ...
+
+    class MissingTopicForChangelog(Exception):
+        ...
+
+    @property
+    def topics(self) -> TopicMap:
+        return self._topics
+
+    @property
+    def topics_list(self) -> TopicList:
+        return dict_values(self._topics)
+
+    @property
+    def changelog_topics(self) -> Dict[str, TopicMap]:
+        return self._changelog_topics
+
+    @property
+    def changelog_topics_list(self) -> TopicList:
+        return dict_values(self._changelog_topics)
+
+    @property
+    def all_topics(self) -> TopicList:
+        return self.topics_list + self.changelog_topics_list
+
+    @property
+    def pretty_formatted_topic_configs(self) -> str:
+        """
+        Returns a print-friendly version of all the topics and their configs
+
+        :return: a pprint-formatted string of all the topics
+        """
+        return pprint.pformat(
+            {
+                topic.name: topic.config.__dict__ if topic.config else "NO CONFIG"
+                for topic in self.all_topics
+            }
+        )
+
+    @property
+    def admin(self) -> Admin:
+        """
+        Raises an exception so that things that require an Admin instance fail.
+        """
+        if not self._admin:
+            raise self.MissingAdmin(
+                "No Admin client has been defined; add one with 'set_admin()'"
+            )
+        return self._admin
+
+    @property
+    def has_admin(self) -> bool:
+        """
+        Whether an admin client has been defined or not.
+
+        :return: bool
+        """
+        return bool(self._admin)
+
+    def set_admin(self, admin: Admin):
+        """
+        Allows for adding an Admin class post-init.
+
+        :param admin: an Admin instance
+        """
+        self._admin = admin
+
+    @abstractmethod
+    def topic_config(
+        self,
+        num_partitions: Optional[int] = None,
+        replication_factor: Optional[int] = None,
+        extra_config: Optional[Mapping] = None,
+    ) -> TopicConfig:
+        """
+        Convenience method for generating a `TopicConfig` with default settings
+
+        :param num_partitions: the number of topic partitions
+        :param replication_factor: the topic replication factor
+        :param extra_config: other optional configuration settings
+
+        :return: a TopicConfig object
+        """
+        ...
+
+    @abstractmethod
+    def topic(
+        self,
+        name: str,
+        value_deserializer: Optional[DeserializerType] = None,
+        key_deserializer: Optional[DeserializerType] = "bytes",
+        value_serializer: Optional[SerializerType] = None,
+        key_serializer: Optional[SerializerType] = "bytes",
+        config: Optional[TopicConfig] = None,
+        auto_create_config: bool = True,
+    ) -> Topic:
+        """
+        A convenience method for generating a `Topic`. Will use default config options
+        as dictated by the TopicManager.
```

Comment:
It was confusing for me wether this method actually creates a topic in kafka or just "sets it up".
In one way I understand why the name is just `topic` (and `changelog_topic` for changelogs) and I am not 100% sure how to improve it but in another hand I feel like the name of a method should be a verb and should describe what it does better. I would maybe change it to `add_topic` to still keep it short. But I am not sure. Leaving this comment to open a discussion

Reply:
This hits on one of the largest pain points of the library in my mind (and my additions compound them).

It can get a little tricky to distinguish between the `Topic` object, which may or may not actually exist in Kafka, and an actual Kafka "topic". Even worse, we readily use `topic` as an argument, which can then only be distinguished as to whether it's a topic name (`str`) or a `Topic` via the type hints, which doesn't feel super good. 

Plus, `changelogs`, which I've now added, are _also_ `Topic`s and run into the same issues with argument names, and are often addressed as a distinct thing for the sake of more easily tracking stuff.

I think we maybe need a better argument or object naming scheme.

Anywho, back to your point. 

I kind of like the `register_topic` name, especially since the `TopicManager` tracks all `Topic`s generated by it so it better captures _that_ behavior. 

However, some may interpret that as creating the topic; one nice thing of _not_ having a verb is I think it helps avoid giving the false impression that it's generating the actual Kafka topic when you make the `Topic` object. It's all subjective, really...as evidenced by you (understandably) not being sure if it created the topic or not. IDK.

I think it might just boil down to better documentation in the docstring?